Source/WebCore/html/HTMLCanvasElement.cpp

```diff
@@ -538,6 +555,49 @@ ImageBitmapRenderingContext* HTMLCanvasElement::getContextBitmapRenderer(const S
     return static_cast<ImageBitmapRenderingContext*>(m_context.get());
 }
 
+bool HTMLCanvasElement::isWebGPUType(const String& type)
+{
+    return type == "webgpu"_s;
+}
+
+GPUCanvasContext* HTMLCanvasElement::createContextWebGPU(const String& type)
+{
+    ASSERT_UNUSED(type, HTMLCanvasElement::isWebGPUType(type));
+    ASSERT(!m_context);
+
+    if (!document().settings().webGPU())
+        return nullptr;
+
+    m_context = GPUCanvasContext::create(*this);
+
+    if (m_context) {
+        // Adds the current document as an observer, so that the document calls prepareForDisplay
+        // when the canvas changes.
+        addObserver(document());
+
+        // Need to make sure a RenderLayer and compositing layer get created for the Canvas.
+        invalidateStyleAndLayerComposition();
+    }
+
+    return static_cast<GPUCanvasContext*>(m_context.get());
+}
+
+GPUCanvasContext* HTMLCanvasElement::getContextWebGPU(const String& type)
+{
+    ASSERT_UNUSED(type, HTMLCanvasElement::isWebGPUType(type));
+
+    if (!document().settings().webGPU())
+        return nullptr;
+
+    if (m_context && !m_context->isWebGPU())
+        return nullptr;
```

Comment:
Maybe an `ASSERT_NOT_REACHED` here since it seems like a very unlikely scenario that should probably be hit in debug mode.

Reply:
If I interpret this correctly, one `<canvas>` should only ever have one context. If we already have a WebGL context for a canvas, then trying to get a WebGPU context should fail gracefully, because we already have a context, but it's a WebGL one, not WebGPU.

Reply:
https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/getContext:

> The HTMLCanvasElement.getContext() method returns a drawing context on the canvas, or null if the context identifier is not supported, or the canvas has already been set to a different context mode.

> Later calls to this method on the same canvas element, with the same contextType argument, will always return the same drawing context instance as was returned the first time the method was invoked. It is not possible to get a different drawing context object on a given canvas element.

Reply:
But isn't that context already checked at the call site? If checking it again yields a different result then there is a major issue.

Reply:
Only place where `::getContextWebGPU` (where that code belongs) is used is `::getContext`, `::getContext` only checks the `type` string and redirects it to the corresponding method. Or maybe I miss somewhere else :/

Reply:
Yup, only the type is checked at those sites you're right.